fix rosbag record

davrsky opened the pull request:
Use the timestamp from the message event, instead of ros::time::now --
this should help fix the problem of getting late timestamps in theory

**ros subscribers:**
In the background, ros creates a thread (or multiple threads?) that accept tcp connections, recieve messages, and store into an internal buffer / queue, one queue per topic
(length of queue set by the queue_size(number of messages) & buffer_size(number of bytes) options on ros subscriber constructor).

Then, the recieved messages are dispatched to subscriber callbacks. this either happens single threaded (if you call spin or spinOnce), or multithreaded (if you use asyncspinner).

**rosbag record:** (which is just a ros node, nothing magic about it)
In rosbag record, it's using asyncspinner w/ 10 threads, and it subscribes to many topics, all to the same callback doQueue, which when called, checks the current time, and adds the message + timestamp onto a queue. 

There's a separate thread in rosbag record that handles reading from that queue and writing the messages sequentially to disk. 

There's a lock that synchronizes between the threads, so that no two threads are writing to the queue at the same time & corrupting it. However this means while writing messages out to disk, new messages can't be added on to the queue, so those threads will pause at the line of code that acquires the lock. Since this is AFTER capturing the timestamp, you'd think this would be ok...but the problem is we only have 10 threads handling callbacks, so if writing messages out to disk takes longer than the time for 10 messages to come in, the messages will be stuck on the internal ros queue BEFORE calling our callback, meaning the ros::Time::now call wont be called until later, getting a later timestamp. We don't actually care about the time the message was moved from the internal ros queue onto the rosbag record queue, what we care about is when the message was actually published. We can't get that directly, so the next best thing is when the message was actually recieved.

davrsky committed:
rosbag record: Fix recording timestamps

Use the timestamp from the message event, instead of ros::time::now --
this should help fix the problem of getting late timestamps in theory

davidbailey commented:
Is there any chance we could get this pushed into the upstream repo, ros/ros_comm, by convincing them this is a better way?

davrsky commented:
https://github.com/ros/ros_comm/pull/1775 trying to merge upstream

davrsky committed:
rosbag record: Increase subscriber queue size

If you're recording a 100hz topic, the default queue size means we can
only tolerate delays up to 1s before we start losing messages.
Increasing the queue size to 1000 should give us a reasonable buffer.

yogeshgstarsky commented on tools/rosbag/src/recorder.cpp:
```diff
@@ -217,7 +217,7 @@ shared_ptr<ros::Subscriber> Recorder::subscribe(string const& topic) {
 
     ros::SubscribeOptions ops;
     ops.topic = topic;
-    ops.queue_size = 100;
```
Is the idea here to check with a larger queue size and see if that fixes the issue?
 If so, should we try `rectime` rather than `msg_event.getReceiptTime()`. It would alse be useful to print the queue sizes frequently during writing to see if it actually gets full!

yogeshgstarsky replied:
Is the idea here to check with a larger queue size and see if that fixes the issue?
If so, should we try rectime rather than msg_event.getReceiptTime(). It would alse be useful to print the queue sizes frequently during writing to see if it actually gets full!

yogeshgstarsky reviewed (commented):
looks like we'll need to investigate this further?

davrsky commented:
There's 2 queues involved, one on each subscriber, and one on the node itself before it gets written out. The latter one it does report an error if the queue maxes out, and I do see this log message occasionally in the starsky logs, but much less often than for every bag, so I don't think this is the main problem. The other queue, in theory should only be queued for long enough to write one message out to disk, but in my stress test it seems to make a difference if I increase that queue.

Both changes seem to improve things in my test case, but we should test it on an actual truck, with actual data, to see how much it improves there.

davrsky committed:
rosbag recorder: Wait for valid time BEFORE subscribing

Otherwise we get into the case where we could capture some messages with
invalid timestamps, which will cause rosbag storage to throw an
exception when we attempt to write messages with invalid timestamps

davrsky committed:
add a note that we're running a modified version of rosbag record
